src/core/recv_buffer.c

```diff
@@ -969,26 +1202,71 @@ QuicRecvBufferDrain(
     )
 {
     CXPLAT_DBG_ASSERT(DrainLength <= RecvBuffer->ReadPendingLength);
+
+    //
+    // Mark chunks as no longer externally referenced and reset the read-pending data length.
+    // For Multiple mode, this is done when each chunk is drained.
+    //
     if (RecvBuffer->RecvMode != QUIC_RECV_BUF_MODE_MULTIPLE) {
+        for (CXPLAT_LIST_ENTRY* Link = RecvBuffer->Chunks.Flink;
+            Link != &RecvBuffer->Chunks;
+            Link = Link->Flink) {
+            QUIC_RECV_CHUNK* Chunk =
+                CXPLAT_CONTAINING_RECORD(Link, QUIC_RECV_CHUNK, Link);
+            Chunk->ExternalReference = FALSE;
```

Comment:
I don't understand why you've changed how/when we set/clear `ExternalReference`.

Reply:
In app-owned mode, on a drain, I need to dereference all chunks that have been referenced, even if not enough data is drained to reach them in the draining loop below.

For Single / Circular mode, what needs to happen is similar: a drain dereference everything. Except that for these modes, the limitation on the number of chunks (1 active and potentially 1 to delete) meant the logic in `QuicRecvBufferPartialDrain` would work.

The drain logic is already the most complex part of recv_buffer, so I am trying to consolidate the dereferencing here for the mode that behave the same in that regard.